service/storage_service.cc

```diff
@@ -1158,6 +1158,46 @@ class topology_coordinator {
         }
     }
 
+    // Preconditions:
+    // - There are no nodes that are trying to join and there are no topology
+    //   operations in progress
+    // - `features_to_enable` represents a set of features that are currently
+    //   marked as supported by all normal nodes and it is not empty
+    future<> enable_features(group0_guard guard, std::set<sstring> features_to_enable) {
+        if (!_topo_sm._topology.new_nodes.empty() || !_topo_sm._topology.transition_nodes.empty()) {
+            on_internal_error(slogger,
+                    "topology coordinator attempted to enable features even though there are"
+                    " joining nodes or topology operations in progress");
+        }
+
+        // If we are here, then we noticed that all nodes support some features
+        // that are not enabled yet. Perform a global barrier to make sure that:
+        //
+        // 1. All nodes saw (and persisted) a view of the system.topology table
+        //    that is equal to what the topology coordinator sees (or newer,
+        //    but in that case updating the topology state will fail),
+        // 2. None of the nodes is restarting at the moment and trying to
+        //    update its corresponding `supported_features` column (that's why
+        //    we use `barrier_after_feature_update` instead of regular `barrier`).
+        //
+        // After we get a successful confirmation from each node, we have
+        // a guarantee that they won't attempt to revoke support for those
+        // features. That's because we do not allow nodes to boot without
+        // a feature that is supported by all nodes in the cluster, even if
+        // the feature is not enabled yet.
+        guard = co_await exec_global_command(std::move(guard),
+                raft_topology_cmd{raft_topology_cmd::command::barrier_after_feature_update},
+                {_raft.id()},
+                drop_guard_and_retake::no);
+
+        topology_mutation_builder builder(guard.write_timestamp());
+        builder.add_enabled_features(features_to_enable);
+        auto reason = ::format("enabling features: {}", features_to_enable);
+        co_await update_topology_state(std::move(guard), {builder.build()}, reason);
```

Comment:
So if we issue the regular barrier while holding the guard and some nodes changes its features set meanwhile this will fail. Hence we do not need the special barrier.

Reply:
Yes, but the special barrier is to protect from a different kind of problem. If you used a regular barrier, you can have a situation where:

- Node A is started, commits a command which adds feature X to its `supported_features` but crashes before applying locally,
- Node A is restarted without feature X and passes the feature check, then starts its server for group 0,
- Topology coordinator sees that it can enable a feature, does a global barrier which succeeds, and then applies a command that adds X to `enabled_features`.

The special barrier is here in order to make sure that the node applied its command that changes its `supported_features` column and won't disable them without checking whether it is safe to do so.

(moved this comment from a different place here)

Reply:
> Yes, but the special barrier is to protect from a different kind of problem. If you used a regular barrier, you can have a situation where:
> 
>     * Node A is started, commits a command which adds feature X to its `supported_features` but crashes before applying locally,
> 
>     * Node A is restarted without feature X and passes the feature check, then starts its server for group 0,

If node A crashed it will have to synchronize the topology first because of the `must_synchronize_topology` flag which is there for exactly this reason. And after that it should see that it starts without a feature that was previously available and will have to check it against enabled features and features that can be enabled.

> 
>     * Topology coordinator sees that it can enable a feature, does a global barrier which succeeds, and then applies a command that adds X to `enabled_features`.
> 
> 
> The special barrier is here in order to make sure that the node applied its command that changes its `supported_features` column and won't disable them without checking whether it is safe to do so.
> 

I do not understand neither the problem nor how this flags fixes it. Please explain it step by step. As of now it looks like it is n attempt to cover some other missing check somewhere.

Reply:
> If node A crashed it will have to synchronize the topology first because of the must_synchronize_topology flag which is there for exactly this reason. And after that it should see that it starts without a feature that was previously available and will have to check it against enabled features and features that can be enabled.

Checking against enabled features happens before starting group 0, and much earlier than `storage_service::update_topology_with_local_metadata`. Adding a check to that function doesn't help because it's too late - the topology coordinator may have already decided to enable a feature this node doesn't understand, and this node might have applied some commands that depend on those features.

> I do not understand neither the problem nor how this flags fixes it. Please explain it step by step. As of now it looks like it is n attempt to cover some other missing check somewhere.

The problem is the race as I described in the previous message.

The topology coordinator does the global barrier to get the confirmation that nodes see the same values of `supported_features` columns as the topology coordinator and won't revoke support for the feature. The special barrier only succeeds after `update_topology_with_local_metadata` was called, so this makes sure that the node won't modify that column until restart. On restart, the node will do the feature check, so this makes sure that it doesn't revoke support for the feature.

Reply:
> > If node A crashed it will have to synchronize the topology first because of the must_synchronize_topology flag which is there for exactly this reason. And after that it should see that it starts without a feature that was previously available and will have to check it against enabled features and features that can be enabled.
> 
> Checking against enabled features happens before starting group 0, and much earlier than `storage_service::update_topology_with_local_metadata`. Adding a check to that function doesn't help because it's too late - the topology coordinator may have already decided to enable a feature this node doesn't understand, and this node might have applied some commands that depend on those features.
> 
> > I do not understand neither the problem nor how this flags fixes it. Please explain it step by step. As of now it looks like it is n attempt to cover some other missing check somewhere.
> 
> The problem is the race as I described in the previous message.
> 
> The topology coordinator does the global barrier to get the confirmation that nodes see the same values of `supported_features` columns as the topology coordinator and won't revoke support for the feature. The special barrier only succeeds after `update_topology_with_local_metadata` was called, so this makes sure that the node won't modify that column until restart. On restart, the node will do the feature check, so this makes sure that it doesn't revoke support for the feature.



> > If node A crashed it will have to synchronize the topology first because of the must_synchronize_topology flag which is there for exactly this reason. And after that it should see that it starts without a feature that was previously available and will have to check it against enabled features and features that can be enabled.
> 
> Checking against enabled features happens before starting group 0, and much earlier than `storage_service::update_topology_with_local_metadata`. Adding a check to that function doesn't help because it's too late - the topology coordinator may have already decided to enable a feature this node doesn't understand, and this node might have applied some commands that depend on those features.

In your scenario a node drops a feature that it previously declared as enabled. This is handled in `storage_service::update_topology_with_local_metadata`. The check si missing there under the group0 lock that this is allowed to do.

> 
> > I do not understand neither the problem nor how this flags fixes it. Please explain it step by step. As of now it looks like it is n attempt to cover some other missing check somewhere.
> 
> The problem is the race as I described in the previous message.
> 
> The topology coordinator does the global barrier to get the confirmation that nodes see the same values of `supported_features` columns as the topology coordinator and won't revoke support for the feature. The special barrier only succeeds after `update_topology_with_local_metadata` was called, so this makes sure that the node won't modify that column until restart. On restart, the node will do the feature check, so this makes sure that it doesn't revoke support for the feature.

A node cannot modify its features while the barrier is runing. If it does either the update of the enabled_features will fail on the coordinator and it will have to recheck again or the update of the supported_features on the node will fail and the node needs to check again that it is allowed to downgrade. All I see is a missing check in `storage_service::update_topology_with_local_metadata`.

Reply:
OK, I'll try to explain the scenario in more detail.

The cluster consists of nodes A, B, C. Node A and B support feature X. Node C does not support X. All of the nodes are up and running. A is the topology coordinator.

Now:

1. Node C is stopped.
2. A new version of Scylla is installed on node C that supports feature X.
3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.
4. Node C is downgraded back to the old version that does not support X.
5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.
6. Node C starts its group 0 server.
7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.
8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it updates its `supported_features` column.

As a result, we have enabled X but it is not supported by C.

The above scenario assumes that we use a regular barrier. If we change it to use the special barrier, the barrier won't succeed until `update_topology_with_local_metadata` is called. So, it makes sure that point 8. always happens before point 7.

Reply:
> OK, I'll try to explain the scenario in more detail.
> 
> The cluster consists of nodes A, B, C. Node A and B support feature X. Node C does not support X. All of the nodes are up and running. A is the topology coordinator.
> 
> Now:
> 
>     1. Node C is stopped.
> 
>     2. A new version of Scylla is installed on node C that supports feature X.
> 
>     3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.
> 
>     4. Node C is downgraded back to the old version that does not support X.
> 
>     5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.
> 
>     6. Node C starts its group 0 server.
> 
>     7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.
> 
>     8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it updates its `supported_features` column.

And this is exactly where you should fix the bug. At this point in `storage_service::update_topology_with_local_metadata` the barrier will be called (because of a previous crash) and  the node C will see that it supported feature before X and it has to check that if all other nodes support it as well it cannot be downgraded and abort the startup. We need this check anyway. Where is it done now?

> 
> 
> As a result, we have enabled X but it is not supported by C.
> 
> The above scenario assumes that we use a regular barrier. If we change it to use the special barrier, the barrier won't succeed until `update_topology_with_local_metadata` is called. So, it makes sure that point 8. always happens before point 7.

We do not need to wait for anything just use mutually exclusion that the raft guard provides.

Reply:
> > OK, I'll try to explain the scenario in more detail.
> > The cluster consists of nodes A, B, C. Node A and B support feature X. Node C does not support X. All of the nodes are up and running. A is the topology coordinator.
> > Now:
> > ```
> > 1. Node C is stopped.
> > 
> > 2. A new version of Scylla is installed on node C that supports feature X.
> > 
> > 3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.
> > 
> > 4. Node C is downgraded back to the old version that does not support X.
> > 
> > 5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.
> > 
> > 6. Node C starts its group 0 server.
> > 
> > 7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.
> > 
> > 8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it updates its `supported_features` column.
> > ```
> 
> And this is exactly where you should fix the bug. At this point in `storage_service::update_topology_with_local_metadata` the barrier will be called (because of a previous crash) and the node C will see that it supported feature before X and it has to check that if all other nodes support it as well it cannot be downgraded and abort the startup. We need this check anyway. Where is it done now?

The feature check is done much earlier than `update_topology_with_local_metadata` - here:

https://github.com/scylladb/scylladb/blob/39ca07c49b257d37856164ea5b845f7b98224b98/main.cc#L1315

This is done pretty early, before commitlog replay or group 0 server start, not to mention `storage_service::join_cluster`.

> 
> > As a result, we have enabled X but it is not supported by C.
> > The above scenario assumes that we use a regular barrier. If we change it to use the special barrier, the barrier won't succeed until `update_topology_with_local_metadata` is called. So, it makes sure that point 8. always happens before point 7.
> 
> We do not need to wait for anything just use mutually exclusion that the raft guard provides.

I'm not sure how we could make it work. There are components that depend on features during initialization, and need to be initialized before we get to that place and start a group 0 operation (commitlog, group 0 server). The features must be stored persistently somewhere where we can read them before joining group 0.

(By the way, I realized that we will also need to flush the topology table when applying a command that changes the supported_features/enabled_features columns for this to work, as we do the check before commitlog replay)

Reply:
> > > OK, I'll try to explain the scenario in more detail.
> > > The cluster consists of nodes A, B, C. Node A and B support feature X. Node C does not support X. All of the nodes are up and running. A is the topology coordinator.
> > > Now:
> > > ```
> > > 1. Node C is stopped.
> > > 
> > > 2. A new version of Scylla is installed on node C that supports feature X.
> > > 
> > > 3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.
> > > 
> > > 4. Node C is downgraded back to the old version that does not support X.
> > > 
> > > 5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.
> > > 
> > > 6. Node C starts its group 0 server.
> > > 
> > > 7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.
> > > 
> > > 8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it updates its `supported_features` column.
> > > ```
> > 
> > 
> > And this is exactly where you should fix the bug. At this point in `storage_service::update_topology_with_local_metadata` the barrier will be called (because of a previous crash) and the node C will see that it supported feature before X and it has to check that if all other nodes support it as well it cannot be downgraded and abort the startup. We need this check anyway. Where is it done now?
> 
> The feature check is done much earlier than `update_topology_with_local_metadata` - here:
> 
> https://github.com/scylladb/scylladb/blob/39ca07c49b257d37856164ea5b845f7b98224b98/main.cc#L1315
> 
> This is done pretty early, before commitlog replay or group 0 server start, not to mention `storage_service::join_cluster`.

As far as I see this only checks that already enabled feature is not dropped, but we also need to check that a feature that is supported by all nodes (but not yet enabled) is also not dropped. This is because it could have been enabled already but this node did not get that info yet. The enablement procedure makes sure that before `enabled_features` is updated all nodes will have up-to-date `supported_feature` for the entire cluster, so if  `enabled_features`  is outdated `supported_feature` are guarantied to not be.

Also it is correct to do it here in the early stage, but this check may work on an outdated information and hence miss some cases. But if it happens dropped features are guarantied to not be enabled yet. This is why the check (or at least part of it) has to be repeated after the read barrier.

> 
> > > As a result, we have enabled X but it is not supported by C.
> > > The above scenario assumes that we use a regular barrier. If we change it to use the special barrier, the barrier won't succeed until `update_topology_with_local_metadata` is called. So, it makes sure that point 8. always happens before point 7.
> > 
> > 
> > We do not need to wait for anything just use mutually exclusion that the raft guard provides.
> 
> I'm not sure how we could make it work. There are components that depend on features during initialization, and need to be initialized before we get to that place and start a group 0 operation (commitlog, group 0 server). The features must be stored persistently somewhere where we can read them before joining group 0.
>

I did not say they should not be stored persistently. I said that there is a missing check for features that cannot be dropped because all nodes declared them as enabled already. If we will have the check we do not need to wait for `update_topology_with_local_metadata`  to complete because in case the feature was dropped it will simply abort the node.

Reply:
> > > > OK, I'll try to explain the scenario in more detail.
> > > > The cluster consists of nodes A, B, C. Node A and B support feature X. Node C does not support X. All of the nodes are up and running. A is the topology coordinator.
> > > > Now:
> > > > ```
> > > > 1. Node C is stopped.
> > > > 
> > > > 2. A new version of Scylla is installed on node C that supports feature X.
> > > > 
> > > > 3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.
> > > > 
> > > > 4. Node C is downgraded back to the old version that does not support X.
> > > > 
> > > > 5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.
> > > > 
> > > > 6. Node C starts its group 0 server.
> > > > 
> > > > 7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.
> > > > 
> > > > 8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it updates its `supported_features` column.
> > > > ```
> > > 
> > > 
> > > And this is exactly where you should fix the bug. At this point in `storage_service::update_topology_with_local_metadata` the barrier will be called (because of a previous crash) and the node C will see that it supported feature before X and it has to check that if all other nodes support it as well it cannot be downgraded and abort the startup. We need this check anyway. Where is it done now?
> > 
> > 
> > The feature check is done much earlier than `update_topology_with_local_metadata` - here:
> > https://github.com/scylladb/scylladb/blob/39ca07c49b257d37856164ea5b845f7b98224b98/main.cc#L1315
> > 
> > This is done pretty early, before commitlog replay or group 0 server start, not to mention `storage_service::join_cluster`.
> 
> As far as I see this only checks that already enabled feature is not dropped, but we also need to check that a feature that is supported by all nodes (but not yet enabled) is also not dropped. This is because it could have been enabled already but this node did not get that info yet.

This PR modifies this function to do that - we now check both against `enabled_features` and the intersection of `supported_features` of all (normal) nodes. So, this is already handled.

> The enablement procedure makes sure that before `enabled_features` is updated all nodes will have up-to-date `supported_feature` for the entire cluster, so if `enabled_features` is outdated `supported_feature` are guarantied to not be.

This is not true if the node crashes after committing a command that modifies its `supported_feature` column, but before applying it. This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass. Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.

Maybe I'm missing something, but I really don't see how the checks you propose prevent this situation. If I'm wrong, then please explain it using my 8-point scenario as an example.

Reply:
> > > > > OK, I'll try to explain the scenario in more detail.
> > > > > The cluster consists of nodes A, B, C. Node A and B support feature X. Node C does not support X. All of the nodes are up and running. A is the topology coordinator.
> > > > > Now:
> > > > > ```
> > > > > 1. Node C is stopped.
> > > > > 
> > > > > 2. A new version of Scylla is installed on node C that supports feature X.
> > > > > 
> > > > > 3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.
> > > > > 
> > > > > 4. Node C is downgraded back to the old version that does not support X.
> > > > > 
> > > > > 5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.
> > > > > 
> > > > > 6. Node C starts its group 0 server.
> > > > > 
> > > > > 7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.
> > > > > 
> > > > > 8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it updates its `supported_features` column.
> > > > > ```
> > > > 
> > > > 
> > > > And this is exactly where you should fix the bug. At this point in `storage_service::update_topology_with_local_metadata` the barrier will be called (because of a previous crash) and the node C will see that it supported feature before X and it has to check that if all other nodes support it as well it cannot be downgraded and abort the startup. We need this check anyway. Where is it done now?
> > > 
> > > 
> > > The feature check is done much earlier than `update_topology_with_local_metadata` - here:
> > > https://github.com/scylladb/scylladb/blob/39ca07c49b257d37856164ea5b845f7b98224b98/main.cc#L1315
> > > 
> > > This is done pretty early, before commitlog replay or group 0 server start, not to mention `storage_service::join_cluster`.
> > 
> > 
> > As far as I see this only checks that already enabled feature is not dropped, but we also need to check that a feature that is supported by all nodes (but not yet enabled) is also not dropped. This is because it could have been enabled already but this node did not get that info yet.
> 
> This PR modifies this function to do that - we now check both against `enabled_features` and the intersection of `supported_features` of all (normal) nodes. So, this is already handled.

Indeed. I was looking at outdated master. So this part is fine, but the check should be done after sync with the group0 as well.

> 
> > The enablement procedure makes sure that before `enabled_features` is updated all nodes will have up-to-date `supported_feature` for the entire cluster, so if `enabled_features` is outdated `supported_feature` are guarantied to not be.
> 
> This is not true if the node crashes after committing a command that modifies its `supported_feature` column, but before applying it. 

How is it not true? In this case `enabled_features` will not contain new features as well. The point is that new features first appear on all nodes in `supported_feature` and only then they appear in `enabled_features`.

> This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.

Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.

> Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.

And if the barrier runs before  `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.

> 
> Maybe I'm missing something, but I really don't see how the checks you propose prevent this situation. If I'm wrong, then please explain it using my 8-point scenario as an example.

I already did in one of the replies which adds the missing downgrade check in step 8. but here it is again.

1. Node C is stopped.

2. A new version of Scylla is installed on node C that supports feature X.

3. Node C is started, it reaches `storage_service::update_topology_with_local_metadata`; it commits a command that modifies its `supported_features` but then crashes before applying the command locally.

4. Node C is downgraded back to the old version that does not support X.

5. Node C is started again. It does the feature check in `feature_service::enable_features_on_startup`: it reads the old `system.topology` state and notices that A and B declares support for X but C does not, so the check passes.

6. Node C starts its group 0 server.

7. Topology coordinator (A) notices that all nodes' `supported_features` column contain X, but it was not put into `enabled_features` yet. It asks nodes B and C to perform a barrier. The barrier succeeds and a command which adds X to `enabled_features` is successfully committed.

8. Node C continues the startup procedure and reaches `storage_service::update_topology_with_local_metadata` again, where it checks its enabled features against what is stored in group0 and sees that it tries to drop a feature that was enabled by all nodes and aborts.

Alternatively this can be done in the barrier itself in step 7, but I think doing it in  `storage_service::update_topology_with_local_metadata` will work as well and will not require any special barrier.

Reply:
> > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> 
> Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> 
> > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> 
> And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.

Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features. This is because applying some group 0 commands could depend on some features being supported.

With the "double check" solution there is a risk that, between the barrier and `update_topology_with_local_metadata`, the node will receive some commands that it can't understand.

Reply:
> > > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> > 
> > 
> > Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> > > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> > 
> > 
> > And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.
> 
> Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features. 

We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.

> This is because applying some group 0 commands could depend on some features being supported. 

But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.

> With the "double check" solution there is a risk that, between the barrier and `update_topology_with_local_metadata`, the node will receive some commands that it can't understand.

If we really want to prevent it we can do the check `if (topology.my_features != my_current_features) abort()`  in the barrier itself like I wrote in "Alternatively..." above.

Reply:
> > > > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> > > 
> > > 
> > > Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> > > > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> > > 
> > > 
> > > And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.
> > 
> > 
> > Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features.
> 
> We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.

Why do you think this is only an optimistic check? Maybe it is now because we don't flush the table, but if we make sure to flush when applying commands that modify `{enabled|supported}_features` then it will not be just optimistic.

After the topology coordinator does a global, special barrier, you have a guarantee that:

- Each node persistently stored the current state of `system.topology`, including the supported features and enabled features columns,
- No node will modify its `supported_features` column until restart.

After that, we have a guarantee that the feature check on restart will not allow the nodes to revoke support for new features.

> 
> > This is because applying some group 0 commands could depend on some features being supported.
> 
> But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.

Yes, but before we reach `update_topology_with_local_metadata` the topology coordinator may do a barrier and enable the new features.

Reply:
> > > > > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> > > > 
> > > > 
> > > > Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> > > > > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> > > > 
> > > > 
> > > > And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.
> > > 
> > > 
> > > Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features.
> > 
> > 
> > We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.
> 
> Why do you think this is only an optimistic check? Maybe it is now because we don't flush the table, but if we make sure to flush when applying commands that modify `{enabled|supported}_features` then it will not be just optimistic.
> 
> After the topology coordinator does a global, special barrier, you have a guarantee that:
> 
>     * Each node persistently stored the current state of `system.topology`, including the supported features and enabled features columns,
> 
>     * No node will modify its `supported_features` column until restart.
> 
> 
> After that, we have a guarantee that the feature check on restart will not allow the nodes to revoke support for new features.
> 
> > > This is because applying some group 0 commands could depend on some features being supported.
> > 
> > 
> > But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.
> 
> Yes, but before we reach `update_topology_with_local_metadata` the topology coordinator may do a barrier and enable the new features.



> > > > > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> > > > 
> > > > 
> > > > Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> > > > > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> > > > 
> > > > 
> > > > And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.
> > > 
> > > 
> > > Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features.
> > 
> > 
> > We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.
> 
> Why do you think this is only an optimistic check? Maybe it is now because we don't flush the table, but if we make sure to flush when applying commands that modify `{enabled|supported}_features` then it will not be just optimistic.

Because of the scenario you described yourself (crashing after sending raft command that updates supported features). Locally supported features may be outdated. Features of other nodes can be outdated as well btw.


> 
> After the topology coordinator does a global, special barrier, you have a guarantee that:
> 
>     * Each node persistently stored the current state of `system.topology`, including the supported features and enabled features columns,
> 
>     * No node will modify its `supported_features` column until restart.

If all nodes declare support for a feature X no node should revoke it ever even after the restart.

> 
> 
> After that, we have a guarantee that the feature check on restart will not allow the nodes to revoke support for new features.
> 
> > > This is because applying some group 0 commands could depend on some features being supported.
> > 
> > 
> > But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.
> 
> Yes, but before we reach `update_topology_with_local_metadata` the topology coordinator may do a barrier and enable the new features.

Not if the barrier does `if (topology.my_features != my_current_features) ` check like I described in the part you dropped from the reply. And `update_topology_with_local_metadata` should not downgrade either if all nodes have a feature enabled already. Other nodes rely on that in their optimistic check.

Reply:
(I put the reply in the wrong box, again...)

> > > > > > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> > > > > 
> > > > > 
> > > > > Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> > > > > > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> > > > > 
> > > > > 
> > > > > And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.
> > > > 
> > > > 
> > > > Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features.
> > > 
> > > 
> > > We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.
> > 
> > 
> > Why do you think this is only an optimistic check? Maybe it is now because we don't flush the table, but if we make sure to flush when applying commands that modify `{enabled|supported}_features` then it will not be just optimistic.
> > After the topology coordinator does a global, special barrier, you have a guarantee that:
> > ```
> > * Each node persistently stored the current state of `system.topology`, including the supported features and enabled features columns,
> > 
> > * No node will modify its `supported_features` column until restart.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > After that, we have a guarantee that the feature check on restart will not allow the nodes to revoke support for new features.
> > > > This is because applying some group 0 commands could depend on some features being supported.
> > > 
> > > 
> > > But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.
> > 
> > 
> > Yes, but before we reach `update_topology_with_local_metadata` the topology coordinator may do a barrier and enable the new features.
> 
> > > > > > This node can downgraded and then restarted, and the first check in `enable_features_on_startup` will pass.
> > > > > 
> > > > > 
> > > > > Yes, because it works on outdated data. But what is guarantied if the check passes is that new features are not enabled yet.
> > > > > > Of course, the topology coordinator still has to do a global barrier to enable the feature - but the barrier may happen and succeed before node reaches `update_topology_with_local_metadata`. The feature will become enabled even though we have a node with a running group 0 server that does not support the feature.
> > > > > 
> > > > > 
> > > > > And if the barrier runs before `update_topology_with_local_metadata` by the time the function is executed the local topology will already have new features for the node and it we do the check at this point we can abort the downgrade. This is why I am saying that the check is missing.
> > > > 
> > > > 
> > > > Ok, I think now I see the source of my confusion. I'm working under the assumption that we do not want to start group 0 server unless we are absolutely sure that the node supports all of the enabled features.
> > > 
> > > 
> > > We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.
> > 
> > 
> > Why do you think this is only an optimistic check? Maybe it is now because we don't flush the table, but if we make sure to flush when applying commands that modify `{enabled|supported}_features` then it will not be just optimistic.
> 
> Because of the scenario you described yourself (crashing after sending raft command that updates supported features). Locally supported features may be outdated. Features of other nodes can be outdated as well btw.

If there is a node for which `supported_features` in the raft state != local supported features, then the topology coordinator will not enable anything. That's because the special barrier only succeeds if it happens after `update_topology_with_local_metadata`. If anything, the check can be too strong because some other node might revoke support because it crashed after applying.

> 
> > After the topology coordinator does a global, special barrier, you have a guarantee that:
> > ```
> > * Each node persistently stored the current state of `system.topology`, including the supported features and enabled features columns,
> > 
> > * No node will modify its `supported_features` column until restart.
> > ```
> 
> If all nodes declare support for a feature X no node should revoke it ever even after the restart.

Ok, if we add the feature check to `update_topology_with_local_metadata` then this property will become true. I'm not sure this is strictly necessary because the topology coordinator requires features to be synchronized by using the special barrier, but if it simplifies reasoning then I agree we can add it.

> 
> > After that, we have a guarantee that the feature check on restart will not allow the nodes to revoke support for new features.
> > > > This is because applying some group 0 commands could depend on some features being supported.
> > > 
> > > 
> > > But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.
> > 
> > 
> > Yes, but before we reach `update_topology_with_local_metadata` the topology coordinator may do a barrier and enable the new features.
> 
> Not if the barrier does `if (topology.my_features != my_current_features) ` check like I described in the part you dropped from the reply. And `update_topology_with_local_metadata` should not downgrade either if all nodes have a feature enabled already. Other nodes rely on that in their optimistic check.

It looks a bit weird to check the features in the general-purpose barrier command, but it's better than crashing the node after some commands that the node doesn't understand are applied, and we could drop the special barrier. Maybe instead of aborting we could just keep failing the barrier request so that the startup procedure fails in a more controlled manner, after the failed check in `update_topology_with_local_metadata`.

Reply:
> > > > 
> > > > We do not (the optimistic check that we have no makes sure of it), but sometimes we are not sure and we have to start it to check what is the up-to-date state. This is the case after your patch as well.
> > > 
> > > 
> > > Why do you think this is only an optimistic check? Maybe it is now because we don't flush the table, but if we make sure to flush when applying commands that modify `{enabled|supported}_features` then it will not be just optimistic.
> > 
> > 
> > Because of the scenario you described yourself (crashing after sending raft command that updates supported features). Locally supported features may be outdated. Features of other nodes can be outdated as well btw.
> 
> If there is a node for which `supported_features` in the raft state != local supported features, then the topology coordinator will not enable anything. That's because the special barrier only succeeds if it happens after `update_topology_with_local_metadata`. If anything, the check can be too strong because some other node might revoke support because it crashed after applying.

It does not mean the check that you have there now works on up-to-date state. Either `supported_features`  can be outdated (because the node crashed before applying the command that changed it), or locally supported features in the raft state can be outdated (because the node crashed before applying the command that changed it). Yes, the procedure is designed in such a way that if the both check passes even using outdated data it is safe to start group0, but it still does not mean that downgrade should be allowed. For that we need to do the barrier and re-check.


> 
> > > After the topology coordinator does a global, special barrier, you have a guarantee that:
> > > ```
> > > * Each node persistently stored the current state of `system.topology`, including the supported features and enabled features columns,
> > > 
> > > * No node will modify its `supported_features` column until restart.
> > > ```
> > 
> > 
> > If all nodes declare support for a feature X no node should revoke it ever even after the restart.
> 
> Ok, if we add the feature check to `update_topology_with_local_metadata` then this property will become true. I'm not sure this is strictly necessary because the topology coordinator requires features to be synchronized by using the special barrier, but if it simplifies reasoning then I agree we can add it.

Think about your example where node C crashes before updating its raft sate with feature X.  Now node B is restarted with previous version. The existing check will fail and the user will be told that no downgrade is longer possible. But then C restarts with previous version and it downgrades. This is inconsistent and hard to reason. We need to preserve the invariant that if all nodes upgraded no downgrade is possible.

> 
> > > After that, we have a guarantee that the feature check on restart will not allow the nodes to revoke support for new features.
> > > > > This is because applying some group 0 commands could depend on some features being supported.
> > > > 
> > > > 
> > > > But when we start group0 after the optimistic check passes we are sure that new features are not enabled yet, so this is fine.
> > > 
> > > 
> > > Yes, but before we reach `update_topology_with_local_metadata` the topology coordinator may do a barrier and enable the new features.
> > 
> > 
> > Not if the barrier does `if (topology.my_features != my_current_features) ` check like I described in the part you dropped from the reply. And `update_topology_with_local_metadata` should not downgrade either if all nodes have a feature enabled already. Other nodes rely on that in their optimistic check.
> 
> It looks a bit weird to check the features in the general-purpose barrier command, but it's better than crashing the node after some commands that the node doesn't understand are applied, and we could drop the special barrier. Maybe instead of aborting we could just keep failing the barrier request so that the startup procedure fails in a more controlled manner, after the failed check in `update_topology_with_local_metadata`.

Failing a barrier with exception is bad interface. Also I do not mean we should do literally abort(), but controlled shutdown of a node with appropriate log message. We can return failure to the coordinator from the barrier, but I would rather it gets a failure from a disconnect after the node is down and will not retry because one node is dead (btw do we check that all nodes are alive before trying to enable a feature?).